### ISSUE TITLE
Share dialog fails if a grantee cannot be found

### ISSUE THREAD

**Issue by bernd** [Graylog2/graylog2-server#9149]
I just ran into the following error while testing #9123. This happened when I opened the share dialog for a dashboard.

The problem is that user `5f85af1f11075861c403dbae` doesn't exist anymore and so it doesn't show up in `available_grantees`.

While this is a consistency issue in the backend, I think the frontend shouldn't fail that hard.

```
Cannot find grantee with id grn::::user:5f85af1f11075861c403dbae in available grantees

Stack Trace:

Error: Cannot find grantee with id grn::::user:5f85af1f11075861c403dbae in available grantees
    at http://127.0.0.1:39765/app.js:78355:17
    at http://127.0.0.1:39765/app.js:93053:46
    at http://127.0.0.1:39765/app.js:92904:56
    at http://127.0.0.1:39765/app.js:93053:36
    at http://127.0.0.1:39765/app.js:92869:35
    at http://127.0.0.1:39765/app.js:91414:16
    at ArrayMapNode.iterate.HashCollisionNode.iterate (http://127.0.0.1:39765/app.js:91763:11)
    at Map.__iterate (http://127.0.0.1:39765/app.js:91412:32)
    at ToKeyedSequence.__iterate (http://127.0.0.1:39765/app.js:92867:25)
    at KeyedIterable.mappedSequence.__iterateUncached (http://127.0.0.1:39765/app.js:93052:23)
    at seqIterate (http://127.0.0.1:39765/app.js:90641:16)
    at KeyedIterable.Seq.__iterate (http://127.0.0.1:39765/app.js:90311:14)
    at ToIndexedSequence.__iterate (http://127.0.0.1:39765/app.js:92904:25)
    at IndexedIterable.mappedSequence.__iterateUncached (http://127.0.0.1:39765/app.js:93052:23)
    at seqIterate (http://127.0.0.1:39765/app.js:90641:16)
    at IndexedIterable.IndexedSeq.__iterate (http://127.0.0.1:39765/app.js:90357:14)
    at IndexedIterable.toArray (http://127.0.0.1:39765/app.js:94295:23)
    at List (http://127.0.0.1:39765/app.js:92102:62)
    at IndexedIterable.toList (http://127.0.0.1:39765/app.js:94363:14)
    at EntityShareState.get (http://127.0.0.1:39765/app.js:78359:10)
    at EntityShareSettings (http://127.0.0.1:39765/app.js:134651:48)
    at renderWithHooks (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:14803:18)
    at mountIndeterminateComponent (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:17482:13)
    at beginWork (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:18596:16)
    at HTMLUnknownElement.callCallback (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:188:14)
    at Object.invokeGuardedCallbackDev (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:237:16)
    at invokeGuardedCallback (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:292:31)
    at beginWork$1 (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:23203:7)
    at performUnitOfWork (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:22157:12)
    at workLoopSync (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:22130:22)
    at performSyncWorkOnRoot (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:21756:9)
    at eval (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:11089:24)
    at unstable_runWithPriority (webpack://__%5Bname%5D/./node_modules/react-dom/node_modules/scheduler/cjs/scheduler.development.js?:653:12)
    at runWithPriority$1 (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:11039:10)
    at flushSyncCallbackQueueImpl (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:11084:7)
    at flushSyncCallbackQueue (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:11072:3)
    at scheduleUpdateOnFiber (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:21199:9)
    at dispatchAction (webpack://__%5Bname%5D/./node_modules/react-dom/cjs/react-dom.development.js?:15660:5)
    at Store.<anonymous> (http://127.0.0.1:39765/app.js:110207:9)
    at EventEmitter.eventHandler (webpack://__%5Bname%5D/./node_modules/reflux-core/lib/PublisherMethods.js?:44:22)
    at EventEmitter.emit (webpack://__%5Bname%5D/./node_modules/eventemitter3/index.js?:116:35)
    at Store.trigger (webpack://__%5Bname%5D/./node_modules/reflux-core/lib/PublisherMethods.js?:113:26)
    at Store._trigger (http://127.0.0.1:39765/app.js:48897:12)
    at Store._handleResponse (http://127.0.0.1:39765/app.js:48887:12)
    at tryCatcher (http://127.0.0.1:39765/app.js:106709:23)
    at Promise._settlePromiseFromHandler (http://127.0.0.1:39765/app.js:104825:31)
    at Promise._settlePromise (http://127.0.0.1:39765/app.js:104882:18)
    at Promise._settlePromise0 (http://127.0.0.1:39765/app.js:104927:10)
    at Promise._settlePromises (http://127.0.0.1:39765/app.js:105007:18)
    at _drainQueueStep (http://127.0.0.1:39765/app.js:101804:12)
    at _drainQueue (http://127.0.0.1:39765/app.js:101797:9)
    at Async._drainQueues (http://127.0.0.1:39765/app.js:101813:5)
    at Async.drainQueues (http://127.0.0.1:39765/app.js:101726:14)

Component Stack:

    in EntityShareSettings (created by EntityShareModal)
    in div (created by ModalBody)
    in ModalBody (created by Modal__Body)
    in Modal__Body (created by BootstrapModalConfirm)
    in div (created by ModalDialog)
    in div (created by ModalDialog)
    in div (created by ModalDialog)
    in ModalDialog (created by Modal)
    in Transition (created by Fade)
    in Fade (created by DialogTransition)
    in DialogTransition (created by Modal)
    in RefHolder (created by Modal)
    in div (created by Modal)
    in Portal (created by Modal)
    in Modal (created by Modal)
    in Modal (created by Modal)
    in Modal (created by BootstrapModalWrapper)
    in BootstrapModalWrapper (created by BootstrapModalConfirm)
    in BootstrapModalConfirm (created by EntityShareModal)
    in EntityShareModal (created by ViewList)
    in ViewList (created by DashboardsPage)
    in div (created by Col)
    in Col (created by DashboardsPage)
    in div (created by Row)
    in Row (created by Row)
    in Row (created by DashboardsPage)
    in span (created by DashboardsPage)
    in DocumentTitle (created by DashboardsPage)
    in DashboardsPage (created by ConnectStoresWrapper[DashboardsPage] stores=dashboards)
    in ConnectStoresWrapper[DashboardsPage] stores=dashboards (created by LoadableComponent)
    in LoadableComponent (created by RouterContext)
    in div (created by Col)
    in Col (created by AppWithoutSearchBar)
    in div (created by Row)
    in Row (created by AppWithoutSearchBar__StyledRow)
    in AppWithoutSearchBar__StyledRow (created by AppWithoutSearchBar)
    in div (created by Grid)
    in Grid (created by AppContentGrid)
    in div (created by AppContentGrid__Container)
    in AppContentGrid__Container (created by AppContentGrid)
    in AppContentGrid (created by AppWithoutSearchBar)
    in AppWithoutSearchBar (created by RouterContext)
    in div (created by AppWithGlobalNotifications)
    in AppWithGlobalNotifications (created by RouterContext)
    in RuntimeErrorBoundary (created by Context.Consumer)
    in ReportedErrorBoundary (created by withRouter(ReportedErrorBoundary))
    in withRouter(ReportedErrorBoundary) (created by Context.Consumer)
    in ScratchpadProvider (created by Context.Consumer)
    in App (created by RouterContext)
    in RouterErrorBoundary (created by RouterContext)
    in RouterContext (created by Router)
    in Router (created by AppRouter)
    in AppRouter (created by LoggedInPage)
    in StreamsProvider (created by ConnectStoresWrapper[StreamsProvider] stores=streams)
    in ConnectStoresWrapper[StreamsProvider] stores=streams (created by LoggedInPage)
    in CurrentUserPreferencesProvider (created by LoggedInPage)
    in CurrentUserProvider (created by LoggedInPage)
    in LoggedInPage (created by LoadableComponent)
    in LoadableComponent (created by AppFacade)
    in AppFacade (created by ConnectStoresWrapper[AppFacade] stores=currentUser,server,sessionId)
    in ConnectStoresWrapper[AppFacade] stores=currentUser,server,sessionId
    in ThemeProvider (created by GraylogThemeProvider)
    in GraylogThemeProvider
    in CustomizationProvider
```

**Comment by kmerz** [comment]
I thought we could easily remove the `throw` and ignore it. But for a couple of reasons it is not so simple. Here is the code:
https://github.com/Graylog2/graylog2-server/blob/master/graylog2-web-interface/src/logic/permissions/EntityShareState.js#L131

1. we could remove the thrown and return undefined, but then we would have to `compact`
2. this would lead into a bunch of `flow` problems
3. It actually is better place to throw here.

When we threw the error we actually found a bug in the underlying backend. We assumed that this `throw` would only `throw` when we hit and error, which we did. Since there is no 'easy' way to show here an error message to the user like I did here:
https://github.com/Graylog2/graylog2-server/pull/9307

I would suggest to keep the `throw`.

**Comment by mpfz0r** [comment]
Couldn't you create a Grantee like this then? 
`      return SelectedGrantee.create(granteeId, 'unknown', 'unknown', roleId);`

**Comment by linuspahl** [comment]
@mpfz0r I like the idea, we just need to ensure that the user is not able to submit a wrong configuration, we could get the grantee type based the the granteeId (which is a GRN). In this case we could do something like:

`return SelectedGrantee.create(granteeId, `${entityId} (${granteeType} not found)`, granteeType, roleId);`

**Comment by kmerz** [comment]
Fixed with #9355
